src/Microsoft.AspNet.Mvc.TagHelpers/Internal/AttributeMatcher.cs

```diff
@@ -13,16 +13,20 @@ namespace Microsoft.AspNet.Mvc.TagHelpers.Internal
     public static class AttributeMatcher
     {
         /// <summary>
-        /// Determines the modes a <see cref="ITagHelper" /> can run in based on which modes have all their required
-        /// attributes present, non null, non empty, and non whitepsace.
+        /// Determines the most effective mode a <see cref="ITagHelper" /> can run in based on which modes have
+        /// all their required attributes present.
         /// </summary>
         /// <typeparam name="TMode">The type representing the <see cref="ITagHelper" />'s modes.</typeparam>
         /// <param name="context">The <see cref="TagHelperContext"/>.</param>
         /// <param name="modeInfos">The modes and their required attributes.</param>
-        /// <returns>The <see cref="ModeMatchResult{TMode}"/>.</returns>
-        public static ModeMatchResult<TMode> DetermineMode<TMode>(
+        /// <param name="compare">A comparer delegate.</param>
+        /// <param name="result">The resulting most effective mode.</param>
+        /// <returns><c>true</c> if a result was found, otherwise <c>false</c>.</returns>
+        public static bool  TryDetermineMode<TMode>(
             TagHelperContext context,
-            IReadOnlyList<ModeAttributes<TMode>> modeInfos)
+            IReadOnlyList<ModeAttributes<TMode>> modeInfos,
+            Func<TMode, TMode, int> compare,
+            out TMode result)
         {
             if (context == null)
```

Comment:
Should be consistent with the added `compare` param, either no `null` checks or all `null` checks (since they can't be null). At first we had the `null` checks because it was public API for a while I believe.